src/client/Client.cc

```diff
@@ -3735,7 +3736,13 @@ void Client::_invalidate_inode_cache(Inode *in, int64_t off, int64_t len)
   if (cct->_conf->client_oc) {
     vector<ObjectExtent> ls;
     Striper::file_to_extents(cct, in->ino, &in->layout, off, len, in->truncate_size, ls);
-    objectcacher->discard_set(&in->oset, ls);
+    // discard the cache state after changes are committed to disk (and to
+    // prevent races w/readahead)
+    auto *on_finish = new FunctionContext(
+      [this, in, ls](int r) {
+	objectcacher->discard_set(&in->oset, ls);
+      });
+    objectcacher->discard_writeback(&in->oset, ls, on_finish);
```

Comment:
no need to pass finish context to discard_writeback().

Reply:
ok

Reply:
else if (bh->is_rx()) {
      // cannot remove bh with in-flight read, but we can ensure the
      // read won't overwrite the discard
      bh->last_read_tid = ++oc->last_read_tid;
      bh->bl.clear();
      bh->set_nocache(true);
      oc->mark_zero(bh);
      return;
    } 

I think it's better using continue instead of return in Object::discard( ... ), we should mark all Rx bh to zero instead of th first one,do you think so?

Reply:
yes, you are right